hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -311,7 +311,13 @@ case class TableParallelize(rowsAndGlobal: IR, nPartitions: Option[Int] = None)
 
   protected[ir] override def execute(ctx: ExecuteContext): TableValue = {
     val hc = HailContext.get
-    val Row(_rows: IndexedSeq[_], globals: Row) = CompileAndEvaluate[Row](ctx, rowsAndGlobal, optimize = false)
+    val (ptype, res) = CompileAndEvaluate._apply(ctx, rowsAndGlobal, optimize = false) match {
+      case Left(()) => (PVoid, ().asInstanceOf[Row])
+      case Right((t, off)) => (t.fields(0).typ, SafeRow(t, off).getAs[Row](0))
```

Comment:
So SafeRow essentially destroys any knowledge of the physical type.  So there's no particular reason to use the physical type returned by EvaluateAndCompile when reconstructing the region value in the parallelized partitions.

The right thing here is to have a `literalPType(t: Type, a: Annotation): PType` that chooses the right physical to use for literal values.  This function is actually on my this week's todo.  If you wait or a day or so, I'll have it and you can use it in this PR.

Reply:
> So SafeRow essentially destroys any knowledge of the physical type.

I think I disagree. Requiredness is a statement about the possible range of values, and SafeRow preserves that notion. What we want here is to take the physical type returned by CompileAndEval and call `PType.canonical(pt)` on it to get the constructible type with the same requiredness/etc.

Reply:
I disagree with your disagreement.  The ptype is an inferred approximation to the actual values.  The ptype could be optional but the actual value produced could be present everywhere.  Once you have the actual value, there is no reason no to compute the most precise possible ptype from the value.

You might be worried about the cost of doing that.  In which case the pvalue you got from compile is a convenient upper bound but not the tightest one.

Finally, I think we'll have even better choices for literal pvalues that don't just involve missingness that can give more efficient representation if you know the actual values: sparse arrays, fixed-length arrays, packed small integers, etc.  That should all go in `literalPType`.

Reply:
I agree with your description of what we're doing, but you're right that I'm worried about the cost of inferring the best possible ptype from traversing literal values. They're often quite big, and traversing java objects is slow!

Reply:
I didn't disagree that literalPType may be the right thing to do here (maybe that should take a `ptypeBound` optional arg and bail out to that if it hits some number of traversed objects?), I was disagreeing with your conclusion about SafeRow obliterating what we knew about the value.

Reply:
I think we have a different perspective on ptypes.  Although it isn't the case yet, most of the content of ptypes will not be related to missingness.  If missingness is driving your intuition on ptypes, it will often be wrong.  A ptype is an implementation of a value representation, and SafeRow is an alternative representation of the value.  So by this definition, it literally discards what the ptype was: the representation.

In the practical matters it seems like we agree.  I mostly think we should never be using SafeRow and we can write a fast scanner that pulls out specific information we want for driving the literal ptype decision process.  The natural thing to write with our current infra would unify a bunch of type objects and it would perform like garbage.

Reply:
I don't think a majority of the value of ptypes will come from specializing by representing a restricted set of values: insert fields, drop fields, struct concatenation, column store representation, array concatenation (matrix table join), unpacking struct elements, array filtering, hash vs tree sets and dicts, etc.  None of these have anything to do with specialized values.

I think the defining example of ptype should be column store struct-of-arrays.  That's actually the case that originally motivated all this stuff.

Reply:
We're on the same page. PTypes currently are two things: (1) code-generation interfaces that let us specialize representations for performance, and (2) a system for tracking information about ranges of values (requiredness, fixed-length, plink diploid call, etc). I think we'll want to split these apart at some point, but probably in the near term they'll stay linked. While I agree that the requiredness/length stuff isn't the most important function of ptypes, it's crucial to make sure we get it right, since ptypes are how this information flows into encodings, which have a significant impact on size and read performance.